File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import deque

def bfs(node, graph, visited):
    q = deque([node])
    visited[node] = True
    while q:
        node = q.popleft()
        for _node in graph[node]:
            if not visited[_node]:
                visited[_node] = True
                q.append(_node)

def solution(n, computers):
    answer = 0
    graph = [[] for _ in range(n)]
    visited = [False] * n
    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1:
                graph[i].append(j)
    for i in range(n):
        if not visited[i]:
            answer += 1
            bfs(i, graph, visited)
    return answer

File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
class Solution {
    boolean[] visited;
    int N, M;
    int[][] computers;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        this.computers = computers;
        N = computers.length;
        visited = new boolean[N];

        for (int i=0;i<N;i++) {
            if (!visited[i]) {
                visited[i] = true;
                answer++;
                dfs(i);
            }
        }
        return answer;
    }
    
    void dfs(int i) {
        for (int j=0;j<N;j++) {
            if (computers[i][j] == 1 && !visited[j]) {
                visited[j] = true;
                dfs(j);
            }
        }
    }
}